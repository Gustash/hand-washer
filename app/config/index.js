import local from './local'
import development from './development'
import production from './production'
import colors from './colors'
import store from './store'

let config = {
  colors,
  storybookEnabled: false,
}

if (__DEV__) {
  config = {
    ...config,
    ...development,
    ...local,
    ...store,
  }
} else {
  config = {
    ...config,
    ...local,
    ...production,
    ...store,
  }
}

export { colors }
export default config
