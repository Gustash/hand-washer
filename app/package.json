{
  "name": "app",
  "//": "This allows us to import from e.g. `app/components/Button` from anywhere"
}
