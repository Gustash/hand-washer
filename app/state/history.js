import { createSlice } from '@reduxjs/toolkit'
import moment from 'moment'

const initialState = {
  storeWashes: true,
  washes: [],
}

export default createSlice({
  name: 'history',
  initialState,
  reducers: {
    logWash(state, action) {
      if (!state.storeWashes) {
        return {
          ...state,
          washes: [],
        }
      }

      const { format = 'DD-MM-YYYY HH:mm', date, ...entry } = action.payload
      const momentDate = moment(date, format)

      if (!momentDate.isValid()) {
        return state
      }

      // Get the biggest existing id and increment it to create a new unique one
      const newId = Math.max(0, ...state.washes.map(({ id }) => id)) + 1

      return {
        ...state,
        washes: [
          ...state.washes,
          {
            ...entry,
            id: newId,
            date: momentDate.format('DD-MM-YYYY HH:mm'),
          },
        ],
      }
    },
    setStoreWashes(state, action) {
      return {
        ...state,
        storeWashes: action.payload,
        washes: action.payload ? state.washes : [],
      }
    },
  },
})
