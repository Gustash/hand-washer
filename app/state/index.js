export { actions } from './slices'
export { default as rootReducer } from './reducer'
