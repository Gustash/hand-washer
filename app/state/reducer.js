import { persistReducer } from 'redux-persist'
import AsyncStorage from '@react-native-community/async-storage'
import { combineReducers } from '@reduxjs/toolkit'
import { reducers } from './slices'

const persistConfig = {
  key: 'root',
  storage: AsyncStorage,
}

const rootReducer = combineReducers(reducers)

export default persistReducer(persistConfig, rootReducer)
