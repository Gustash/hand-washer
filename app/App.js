import 'react-native-gesture-handler'
import React from 'react'
import { StyleSheet, StatusBar } from 'react-native'
import { Provider } from 'react-redux'
import { PersistGate } from 'redux-persist/integration/react'
import { NavigationContainer } from '@react-navigation/native'
import { createStackNavigator } from '@react-navigation/stack'
import FontAwesome5 from 'react-native-vector-icons/FontAwesome5'

import Home from './screens/Home'
import WashTimer from './screens/WashTimer'
import Settings from './screens/Settings'
import config, { colors } from './config'

const Stack = createStackNavigator()

const { store, persistor } = config.createStore()

const App = () => (
  <Provider store={store}>
    <PersistGate loading={null} persistor={persistor}>
      <NavigationContainer>
        <StatusBar backgroundColor={colors.green} />
        <Stack.Navigator
          screenOptions={{
            headerStyle: styles.header,
            headerBackImage: () => (
              <FontAwesome5
                name="chevron-left"
                size={16}
                color={colors.white}
                style={styles.headerBackIcon}
              />
            ),
            headerTitleStyle: styles.headerTitle,
            headerRightContainerStyle: styles.headerRightIcon,
            headerBackTitle: 'Back',
            headerBackTitleStyle: [styles.headerTitle, styles.headerBackTitle],
          }}
        >
          <Stack.Screen
            name="Home"
            component={Home}
            options={{ title: 'Wash Timer' }}
          />
          <Stack.Screen name="Settings" component={Settings} />
          <Stack.Screen name="Timer" component={WashTimer} />
        </Stack.Navigator>
      </NavigationContainer>
    </PersistGate>
  </Provider>
)

const styles = StyleSheet.create({
  header: {
    backgroundColor: colors.green,
  },
  headerTitle: {
    color: colors.white,
    fontSize: 17,
  },
  headerRightIcon: {
    marginRight: 16,
  },
  headerBackTitle: {
    fontSize: 16,
  },
  headerBackIcon: {
    marginLeft: 16,
    marginRight: 4,
  },
})

export default App
