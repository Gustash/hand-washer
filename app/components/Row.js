import React from 'react'
import { View, Text, StyleSheet } from 'react-native'
import { colors } from 'app/config'

const Row = ({
  highlight,
  highlightStyle,
  primaryText,
  secondaryText,
  style,
  separator = false,
}) => (
  <View style={[styles.container, separator && styles.separator, style]}>
    <View
      testId="row-highlight-container"
      style={[styles.highlightContainer, highlightStyle]}
    >
      <Text testId="row-highlight-text" style={styles.highlightText}>
        {highlight}
      </Text>
    </View>
    <View style={styles.textContainer}>
      <Text testId="row-primary-text" style={styles.primaryText}>
        {primaryText}
      </Text>
      <Text testId="row-secondary-text" style={styles.secondaryText}>
        {secondaryText}
      </Text>
    </View>
  </View>
)

const styles = StyleSheet.create({
  container: {
    flexDirection: 'row',
    paddingVertical: 16,
  },
  separator: {
    borderBottomColor: colors.separator,
    borderBottomWidth: 1,
  },
  highlightContainer: {
    backgroundColor: colors.green,
    paddingHorizontal: 14,
    paddingVertical: 8,
    width: 80,
    justifyContent: 'center',
    marginHorizontal: 8,
  },
  highlightText: {
    textAlign: 'center',
    color: colors.white,
    fontSize: 17,
  },
  primaryText: {
    fontSize: 17,
    color: colors.darkGrey,
  },
  textContainer: {
    flex: 1,
  },
  secondaryText: {
    fontSize: 14,
    color: colors.lightGrey,
  },
})

export default Row
