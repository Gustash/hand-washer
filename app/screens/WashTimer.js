import React from 'react'
import { ScrollView, StyleSheet } from 'react-native'
import moment from 'moment'
import { useDispatch } from 'react-redux'

import Row from 'app/components/Row'
import Timer from 'app/components/Timer'
import { actions } from 'app/state'
import { colors } from 'app/config'

const WashTimer = () => {
  const dispatch = useDispatch()

  return (
    <ScrollView contentContainerStyle={styles.container}>
      <Timer
        testId="wash-timer-timer"
        duration={20 * 1000}
        style={styles.timer}
        onTimerFinished={() => {
          dispatch(
            actions.logWash({ date: moment().format('DD-MM-YYYY HH:mm') })
          )
        }}
      />
      <Row
        primaryText="Wet your hands"
        secondaryText="With clean, running water (warm or cold), turn off the tap, and apply soap"
        highlight="1"
        highlightStyle={styles.highlightStyle}
        style={styles.row}
      />
      <Row
        primaryText="Lather your hands"
        secondaryText="Rub them together with soap. Lather the backs of your hands, between your fingers, and under your nails"
        highlight="2"
        highlightStyle={styles.highlightStyle}
        style={styles.row}
      />
      <Row
        primaryText="Scrub"
        secondaryText="Scrub your hands for at least 20 seconds"
        highlight="3"
        highlightStyle={styles.highlightStyle}
        style={styles.row}
      />
      <Row
        primaryText="Rinse well"
        secondaryText="Using running clean water"
        highlight="4"
        highlightStyle={styles.highlightStyle}
        style={styles.row}
      />
      <Row
        primaryText="Dry your hands"
        secondaryText="Using a clean towel or air dry them"
        highlight="5"
        highlightStyle={styles.highlightStyle}
        style={styles.row}
      />
    </ScrollView>
  )
}

const styles = StyleSheet.create({
  container: {
    alignItems: 'center',
  },
  timerText: {
    fontSize: 48,
    color: colors.white,
  },
  timer: {
    marginVertical: 80,
  },
  highlightStyle: {
    width: 38,
    height: 36,
  },
  row: {
    marginRight: 65,
    marginLeft: 30,
    paddingVertical: 0,
    marginTop: 16,
  },
})

export default WashTimer
