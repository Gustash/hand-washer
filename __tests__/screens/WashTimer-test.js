import React from 'react'
import { Provider, useDispatch } from 'react-redux'
import configureStore from 'redux-mock-store'
import TestRenderer from 'react-test-renderer'

import WashTimer from '../../app/screens/WashTimer'
import { actions } from '../../app/state'
import Row from '../../app/components/Row'

const mockStore = configureStore([])

const mockDispatch = jest.fn()

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useSelector: jest.fn(),
  useDispatch: jest.fn(),
}))

describe('<WashTimer />', () => {
  let store
  let wrapper

  beforeEach(() => {
    mockDispatch.mockReset()

    global.Date.now = jest.fn(() => new Date('1990-01-01T00:00:00Z').getTime())

    store = mockStore({
      history: {
        storeWashes: true,
        washes: [],
      },
    })

    useDispatch.mockImplementation(() => mockDispatch)

    wrapper = TestRenderer.create(
      <Provider store={store}>
        <WashTimer />
      </Provider>
    )
  })

  it('matches snapshot', () => {
    expect(wrapper.toJSON()).toMatchSnapshot()
  })

  it('logs wash when timer finishes', () => {
    const timer = wrapper.root.findByProps({
      testId: 'wash-timer-timer',
    })
    timer.props.onTimerFinished()
    expect(mockDispatch).toHaveBeenCalledWith(
      actions.logWash({ date: '01-01-1990 00:00' })
    )
  })
})
