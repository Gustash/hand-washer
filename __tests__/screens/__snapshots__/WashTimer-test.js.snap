// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<WashTimer /> matches snapshot 1`] = `
<RCTScrollView
  contentContainerStyle={
    Object {
      "alignItems": "center",
    }
  }
>
  <View>
    <View
      style={
        Object {
          "marginVertical": 80,
        }
      }
    >
      <RNSVGSvgView
        bbHeight={222}
        bbWidth={222}
        focusable={false}
        height={222}
        style={
          Array [
            Object {
              "backgroundColor": "transparent",
              "borderWidth": 0,
            },
            Object {
              "transform": Array [
                Object {
                  "rotateZ": "270deg",
                },
              ],
            },
            Object {
              "flex": 0,
              "height": 222,
              "width": 222,
            },
          ]
        }
        width={222}
      >
        <RNSVGGroup
          matrix={
            Array [
              1,
              0,
              0,
              1,
              0,
              0,
            ]
          }
        >
          <RNSVGCircle
            cx={111}
            cy={111}
            fill={null}
            propList={
              Array [
                "fill",
                "stroke",
                "strokeWidth",
                "strokeDasharray",
                "strokeDashoffset",
                "strokeLinecap",
              ]
            }
            r={106.5}
            stroke={4278232952}
            strokeDasharray={
              Array [
                "669.1592352146259",
                "669.1592352146259",
              ]
            }
            strokeDashoffset={0}
            strokeLinecap={1}
            strokeWidth={9}
          />
        </RNSVGGroup>
      </RNSVGSvgView>
      <View
        accessible={true}
        focusable={false}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "backgroundColor": "#00A1D2",
            "borderRadius": 102,
            "elevation": 16,
            "height": 204,
            "left": 9,
            "opacity": 1,
            "padding": 30,
            "position": "absolute",
            "shadowColor": "#000000",
            "shadowOffset": Object {
              "height": 2,
            },
            "shadowOpacity": 0.2,
            "shadowRadius": 16,
            "top": 9,
            "width": 204,
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "flex": 1,
              "justifyContent": "center",
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#FFFFFF",
                "fontSize": 48,
              }
            }
          >
            20
          </Text>
        </View>
        <Text
          style={
            Object {
              "color": "#FFFFFF",
              "fontSize": 21,
              "textAlign": "center",
            }
          }
        >
          Seconds
        </Text>
      </View>
    </View>
    <View
      style={
        Array [
          Object {
            "flexDirection": "row",
            "paddingVertical": 16,
          },
          false,
          Object {
            "marginLeft": 30,
            "marginRight": 65,
            "marginTop": 16,
            "paddingVertical": 0,
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "backgroundColor": "#00A778",
              "justifyContent": "center",
              "marginHorizontal": 8,
              "paddingHorizontal": 14,
              "paddingVertical": 8,
              "width": 80,
            },
            Object {
              "height": 36,
              "width": 38,
            },
          ]
        }
        testId="row-highlight-container"
      >
        <Text
          style={
            Object {
              "color": "#FFFFFF",
              "fontSize": 17,
              "textAlign": "center",
            }
          }
          testId="row-highlight-text"
        >
          1
        </Text>
      </View>
      <View
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#4A4A4A",
              "fontSize": 17,
            }
          }
          testId="row-primary-text"
        >
          Wet your hands
        </Text>
        <Text
          style={
            Object {
              "color": "#9B9B9B",
              "fontSize": 14,
            }
          }
          testId="row-secondary-text"
        >
          With clean, running water (warm or cold), turn off the tap, and apply soap
        </Text>
      </View>
    </View>
    <View
      style={
        Array [
          Object {
            "flexDirection": "row",
            "paddingVertical": 16,
          },
          false,
          Object {
            "marginLeft": 30,
            "marginRight": 65,
            "marginTop": 16,
            "paddingVertical": 0,
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "backgroundColor": "#00A778",
              "justifyContent": "center",
              "marginHorizontal": 8,
              "paddingHorizontal": 14,
              "paddingVertical": 8,
              "width": 80,
            },
            Object {
              "height": 36,
              "width": 38,
            },
          ]
        }
        testId="row-highlight-container"
      >
        <Text
          style={
            Object {
              "color": "#FFFFFF",
              "fontSize": 17,
              "textAlign": "center",
            }
          }
          testId="row-highlight-text"
        >
          2
        </Text>
      </View>
      <View
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#4A4A4A",
              "fontSize": 17,
            }
          }
          testId="row-primary-text"
        >
          Lather your hands
        </Text>
        <Text
          style={
            Object {
              "color": "#9B9B9B",
              "fontSize": 14,
            }
          }
          testId="row-secondary-text"
        >
          Rub them together with soap. Lather the backs of your hands, between your fingers, and under your nails
        </Text>
      </View>
    </View>
    <View
      style={
        Array [
          Object {
            "flexDirection": "row",
            "paddingVertical": 16,
          },
          false,
          Object {
            "marginLeft": 30,
            "marginRight": 65,
            "marginTop": 16,
            "paddingVertical": 0,
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "backgroundColor": "#00A778",
              "justifyContent": "center",
              "marginHorizontal": 8,
              "paddingHorizontal": 14,
              "paddingVertical": 8,
              "width": 80,
            },
            Object {
              "height": 36,
              "width": 38,
            },
          ]
        }
        testId="row-highlight-container"
      >
        <Text
          style={
            Object {
              "color": "#FFFFFF",
              "fontSize": 17,
              "textAlign": "center",
            }
          }
          testId="row-highlight-text"
        >
          3
        </Text>
      </View>
      <View
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#4A4A4A",
              "fontSize": 17,
            }
          }
          testId="row-primary-text"
        >
          Scrub
        </Text>
        <Text
          style={
            Object {
              "color": "#9B9B9B",
              "fontSize": 14,
            }
          }
          testId="row-secondary-text"
        >
          Scrub your hands for at least 20 seconds
        </Text>
      </View>
    </View>
    <View
      style={
        Array [
          Object {
            "flexDirection": "row",
            "paddingVertical": 16,
          },
          false,
          Object {
            "marginLeft": 30,
            "marginRight": 65,
            "marginTop": 16,
            "paddingVertical": 0,
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "backgroundColor": "#00A778",
              "justifyContent": "center",
              "marginHorizontal": 8,
              "paddingHorizontal": 14,
              "paddingVertical": 8,
              "width": 80,
            },
            Object {
              "height": 36,
              "width": 38,
            },
          ]
        }
        testId="row-highlight-container"
      >
        <Text
          style={
            Object {
              "color": "#FFFFFF",
              "fontSize": 17,
              "textAlign": "center",
            }
          }
          testId="row-highlight-text"
        >
          4
        </Text>
      </View>
      <View
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#4A4A4A",
              "fontSize": 17,
            }
          }
          testId="row-primary-text"
        >
          Rinse well
        </Text>
        <Text
          style={
            Object {
              "color": "#9B9B9B",
              "fontSize": 14,
            }
          }
          testId="row-secondary-text"
        >
          Using running clean water
        </Text>
      </View>
    </View>
    <View
      style={
        Array [
          Object {
            "flexDirection": "row",
            "paddingVertical": 16,
          },
          false,
          Object {
            "marginLeft": 30,
            "marginRight": 65,
            "marginTop": 16,
            "paddingVertical": 0,
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "backgroundColor": "#00A778",
              "justifyContent": "center",
              "marginHorizontal": 8,
              "paddingHorizontal": 14,
              "paddingVertical": 8,
              "width": 80,
            },
            Object {
              "height": 36,
              "width": 38,
            },
          ]
        }
        testId="row-highlight-container"
      >
        <Text
          style={
            Object {
              "color": "#FFFFFF",
              "fontSize": 17,
              "textAlign": "center",
            }
          }
          testId="row-highlight-text"
        >
          5
        </Text>
      </View>
      <View
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#4A4A4A",
              "fontSize": 17,
            }
          }
          testId="row-primary-text"
        >
          Dry your hands
        </Text>
        <Text
          style={
            Object {
              "color": "#9B9B9B",
              "fontSize": 14,
            }
          }
          testId="row-secondary-text"
        >
          Using a clean towel or air dry them
        </Text>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
